### ISSUE TITLE
Feature: publish crate!

### ISSUE THREAD

**Issue by dbkaplun** [KyleMayes/crtq#1]
Hello,

First of all, thanks for making this excellent multi-producer multi-consumer wait-free queue with wait-free memory reclamation!

I would love to use it. Can you please publish this crate?

Thanks for considering!

**Comment by KyleMayes** [comment]
The primary reason I never published this crate is that I was never able to replicate the improved latencies from the CRTurnQueue paper. Judging by the graphs from the paper, the advantages of the CRTurnQueue being wait-free instead of lock-free don't become significant until you have something like 8+ threads competing for access to the queue, and before then you will see significantly worse performance everywhere, and even after that point the CRTurnQueue will still have worse average performance.

For example, here are throughput and latency benchmarks for the unbounded lock-free MPMC queue from my [npnc](https://github.com/KyleMayes/npnc) crate (a Michael-Scott queue using hazard pointers, the same queue the CRTurnQueue is compared to in the paper):

```
1 producer / 1 consumer (two simultaneous threads)

bench throughput_unbounded_mpmc ...
  13_498_419.65 operation/second
bench latency_unbounded_mpmc ...
  consume
    50%:       60.00ns
    70%:       77.00ns
    90%:       82.00ns
    95%:       88.00ns
    99%:       130.00ns
    99.9%:     19_211.00ns
    99.99%:    23_877.00ns
    99.999%:   46_451.00ns
    99.9999%:  201_974.00ns
    99.99999%: 535_811.00ns
  produce
    50%:       67.00ns
    70%:       83.00ns
    90%:       90.00ns
    95%:       108.00ns
    99%:       831.00ns
    99.9%:     1_668.00ns
    99.99%:    13_103.00ns
    99.999%:   15_799.00ns
    99.9999%:  136_735.00ns
    99.99999%: 387_766.00ns
```

And here are the equivalent benchmarks for this crate:

```
1 producer / 1 consumer (two simultaneous threads)

bench throughput_crtq ...
  6_803_091.70 operations/second
bench latency_crtq ...
  consume
    50%:       139.00ns
    70%:       211.00ns
    90%:       251.00ns
    95%:       273.00ns
    99%:       325.00ns
    99.9%:     23_226.00ns
    99.99%:    27_466.00ns
    99.999%:   54_978.00ns
    99.9999%:  210_773.00ns
    99.99999%: 887_854.00ns
  produce
    50%:       120.00ns
    70%:       168.00ns
    90%:       208.00ns
    95%:       222.00ns
    99%:       982.00ns
    99.9%:     1_624.00ns
    99.99%:    13_334.00ns
    99.999%:   15_514.00ns
    99.9999%:  18_629.00ns
    99.99999%: 182_979.00ns
```

The benchmarks for the Michael-Scott queue are very stable, I get roughly the same results every time I run them. The CRTurnQueue latency benchmarks are very volatile, but the above is pretty typical. Average performance is halved when compared to the Michael-Scott queue, with the extremes (99.999%+) showing much worse `consume` times and much better `produce` times.

This is of course with only two threads, I don't have a machine with enough cores to run benchmarks with more than that.

Does my implementation of the CRTurnQueue show much better latencies when there are more threads? Is my implementation wrong? Are my benchmarks bad? Does my implementation of the CRTurnQueue even work when there are more threads? Do I have any idea what I'm doing?

I don't know.

**Comment by dbkaplun** [comment]
I can confirm. I wonder why the p99 latencies for the reference implementation are so high.

Here is the latency benchmark output for the [reference implementation](https://github.com/pramalhe/ConcurrencyFreaks/tree/master/CPP/papers/crturnqueue) on my i7-4650U:
```
$ make latency && ./latency

----- Burst Latency   numThreads=1   kLatencyMeasures=200M -----
##### MichaelScottQueue #####
Aggregating delays for 200 million measurements...
Sorting delays...
Enqueue delay (us): 50%=0  70%=0  80%=0  90%=1  99%=1  99.9%=3  99.99%=12  99.999%=127  max=27584
Dequeue delay (us): 50%=0  70%=0  80%=1  90%=1  99%=1  99.9%=2  99.99%=10  99.999%=337  max=21856
Enqueue delay (us):
50, 0
90, 1
99, 1
99.9, 3
99.99, 12
99.999, 127
Dequeue delay (us):
50, 0
90, 1
99, 1
99.9, 2
99.99, 10
99.999, 337

----- Burst Latency   numThreads=2   kLatencyMeasures=200M -----
##### MichaelScottQueue #####
Aggregating delays for 200 million measurements...
Sorting delays...
Enqueue delay (us): 50%=0  70%=0  80%=1  90%=1  99%=1  99.9%=4  99.99%=49  99.999%=457  max=101381
Dequeue delay (us): 50%=0  70%=1  80%=1  90%=1  99%=3  99.9%=9  99.99%=117  99.999%=1016  max=122053
Enqueue delay (us):
50, 0
90, 1
99, 1
99.9, 4
99.99, 49
99.999, 457
Dequeue delay (us):
50, 0
90, 1
99, 3
99.9, 9
99.99, 117
99.999, 1016

----- Burst Latency   numThreads=4   kLatencyMeasures=200M -----
##### MichaelScottQueue #####
Aggregating delays for 200 million measurements...
Sorting delays...
Enqueue delay (us): 50%=0  70%=1  80%=1  90%=1  99%=2  99.9%=4  99.99%=152  99.999%=3680  max=133606
Dequeue delay (us): 50%=1  70%=1  80%=1  90%=2  99%=7  99.9%=19  99.99%=863  99.999%=5949  max=239839
Enqueue delay (us):
50, 0
90, 1
99, 2
99.9, 4
99.99, 152
99.999, 3680
Dequeue delay (us):
50, 1
90, 2
99, 7
99.9, 19
99.99, 863
99.999, 5949

----- Burst Latency   numThreads=8   kLatencyMeasures=200M -----
##### MichaelScottQueue #####
Aggregating delays for 200 million measurements...
Sorting delays...
Enqueue delay (us): 50%=0  70%=1  80%=1  90%=1  99%=1  99.9%=4  99.99%=181  99.999%=17996  max=207024
Dequeue delay (us): 50%=1  70%=1  80%=1  90%=2  99%=6  99.9%=18  99.99%=3413  99.999%=20935  max=187675
Enqueue delay (us):
50, 0
90, 1
99, 1
99.9, 4
99.99, 181
99.999, 17996
Dequeue delay (us):
50, 1
90, 2
99, 6
99.9, 18
99.99, 3413
99.999, 20935

----- Burst Latency   numThreads=12   kLatencyMeasures=200M -----
##### MichaelScottQueue #####
Aggregating delays for 200 million measurements...
Sorting delays...
Enqueue delay (us): 50%=0  70%=1  80%=1  90%=1  99%=2  99.9%=3  99.99%=188  99.999%=27198  max=165549
Dequeue delay (us): 50%=1  70%=1  80%=1  90%=2  99%=6  99.9%=19  99.99%=5355  99.999%=35168  max=247255
Enqueue delay (us):
50, 0
90, 1
99, 2
99.9, 3
99.99, 188
99.999, 27198
Dequeue delay (us):
50, 1
90, 2
99, 6
99.9, 19
99.99, 5355
99.999, 35168

----- Burst Latency   numThreads=16   kLatencyMeasures=200M -----
##### MichaelScottQueue #####
Aggregating delays for 200 million measurements...
Sorting delays...
Enqueue delay (us): 50%=0  70%=1  80%=1  90%=1  99%=1  99.9%=2  99.99%=114  99.999%=37841  max=190627
Dequeue delay (us): 50%=1  70%=1  80%=1  90%=2  99%=6  99.9%=14  99.99%=8134  99.999%=43782  max=136942
Enqueue delay (us):
50, 0
90, 1
99, 1
99.9, 2
99.99, 114
99.999, 37841
Dequeue delay (us):
50, 1
90, 2
99, 6
99.9, 14
99.99, 8134
99.999, 43782

----- Burst Latency   numThreads=20   kLatencyMeasures=200M -----
##### MichaelScottQueue #####
Aggregating delays for 200 million measurements...
Sorting delays...
Enqueue delay (us): 50%=0  70%=1  80%=1  90%=1  99%=1  99.9%=3  99.99%=187  99.999%=47511  max=179155
Dequeue delay (us): 50%=1  70%=1  80%=1  90%=1  99%=6  99.9%=21  99.99%=10386  99.999%=52524  max=198536
Enqueue delay (us):
50, 0
90, 1
99, 1
99.9, 3
99.99, 187
99.999, 47511
Dequeue delay (us):
50, 1
90, 1
99, 6
99.9, 21
99.99, 10386
99.999, 52524

----- Burst Latency   numThreads=24   kLatencyMeasures=200M -----
##### MichaelScottQueue #####
Aggregating delays for 200 million measurements...
Sorting delays...
Enqueue delay (us): 50%=0  70%=1  80%=1  90%=1  99%=1  99.9%=3  99.99%=160  99.999%=52727  max=212599
Dequeue delay (us): 50%=1  70%=1  80%=1  90%=2  99%=6  99.9%=22  99.99%=12711  99.999%=60342  max=240530
Enqueue delay (us):
50, 0
90, 1
99, 1
99.9, 3
99.99, 160
99.999, 52727
Dequeue delay (us):
50, 1
90, 2
99, 6
99.9, 22
99.99, 12711
99.999, 60342

----- Burst Latency   numThreads=28   kLatencyMeasures=200M -----
##### MichaelScottQueue #####
Aggregating delays for 200 million measurements...
Sorting delays...
Enqueue delay (us): 50%=0  70%=1  80%=1  90%=1  99%=2  99.9%=3  99.99%=159  99.999%=55004  max=184750
Dequeue delay (us): 50%=1  70%=1  80%=1  90%=2  99%=6  99.9%=24  99.99%=14268  99.999%=63761  max=214018
Enqueue delay (us):
50, 0
90, 1
99, 2
99.9, 3
99.99, 159
99.999, 55004
Dequeue delay (us):
50, 1
90, 2
99, 6
99.9, 24
99.99, 14268
99.999, 63761

----- Burst Latency   numThreads=30   kLatencyMeasures=200M -----
##### MichaelScottQueue #####
Aggregating delays for 200 million measurements...
Sorting delays...
Enqueue delay (us): 50%=0  70%=1  80%=1  90%=1  99%=1  99.9%=2  99.99%=138  99.999%=58057  max=193030
Dequeue delay (us): 50%=1  70%=1  80%=1  90%=2  99%=6  99.9%=18  99.99%=16904  99.999%=64394  max=228517
Enqueue delay (us):
50, 0
90, 1
99, 1
99.9, 2
99.99, 138
99.999, 58057
Dequeue delay (us):
50, 1
90, 2
99, 6
99.9, 18
99.99, 16904
99.999, 64394

----- Burst Latency   numThreads=32   kLatencyMeasures=200M -----
##### MichaelScottQueue #####
Aggregating delays for 200 million measurements...
Sorting delays...
Enqueue delay (us): 50%=0  70%=1  80%=1  90%=1  99%=1  99.9%=2  99.99%=166  99.999%=58594  max=212470
Dequeue delay (us): 50%=1  70%=1  80%=1  90%=2  99%=6  99.9%=28  99.99%=18228  99.999%=70382  max=302003
Enqueue delay (us):
50, 0
90, 1
99, 1
99.9, 2
99.99, 166
99.999, 58594
Dequeue delay (us):
50, 1
90, 2
99, 6
99.9, 28
99.99, 18228
99.999, 70382

----- Burst Latency   numThreads=1   kLatencyMeasures=200M -----
##### CRTurnQueue #####
Aggregating delays for 200 million measurements...
Sorting delays...
Enqueue delay (us): 50%=0  70%=0  80%=0  90%=1  99%=1  99.9%=4  99.99%=15  99.999%=112  max=20511
Dequeue delay (us): 50%=0  70%=0  80%=1  90%=1  99%=1  99.9%=3  99.99%=12  99.999%=284  max=8067
Enqueue delay (us):
50, 0
90, 1
99, 1
99.9, 4
99.99, 15
99.999, 112
Dequeue delay (us):
50, 0
90, 1
99, 1
99.9, 3
99.99, 12
99.999, 284

----- Burst Latency   numThreads=2   kLatencyMeasures=200M -----
##### CRTurnQueue #####
Aggregating delays for 200 million measurements...
Sorting delays...
Enqueue delay (us): 50%=0  70%=1  80%=1  90%=1  99%=1  99.9%=5  99.99%=41  99.999%=335  max=43092
Dequeue delay (us): 50%=1  70%=1  80%=1  90%=1  99%=3  99.9%=6  99.99%=83  99.999%=491  max=87279
Enqueue delay (us):
50, 0
90, 1
99, 1
99.9, 5
99.99, 41
99.999, 335
Dequeue delay (us):
50, 1
90, 1
99, 3
99.9, 6
99.99, 83
99.999, 491

----- Burst Latency   numThreads=4   kLatencyMeasures=200M -----
##### CRTurnQueue #####
Aggregating delays for 200 million measurements...
Sorting delays...
Enqueue delay (us): 50%=1  70%=1  80%=1  90%=1  99%=2  99.9%=5  99.99%=224  99.999%=4455  max=132121
Dequeue delay (us): 50%=1  70%=1  80%=1  90%=1  99%=5  99.9%=12  99.99%=433  99.999%=4435  max=57451
Enqueue delay (us):
50, 1
90, 1
99, 2
99.9, 5
99.99, 224
99.999, 4455
Dequeue delay (us):
50, 1
90, 1
99, 5
99.9, 12
99.99, 433
99.999, 4435

----- Burst Latency   numThreads=8   kLatencyMeasures=200M -----
##### CRTurnQueue #####
Aggregating delays for 200 million measurements...
Sorting delays...
Enqueue delay (us): 50%=1  70%=1  80%=1  90%=1  99%=2  99.9%=5  99.99%=541  99.999%=24529  max=155753
Dequeue delay (us): 50%=1  70%=1  80%=1  90%=1  99%=5  99.9%=12  99.99%=3695  99.999%=18728  max=92938
Enqueue delay (us):
50, 1
90, 1
99, 2
99.9, 5
99.99, 541
99.999, 24529
Dequeue delay (us):
50, 1
90, 1
99, 5
99.9, 12
99.99, 3695
99.999, 18728

----- Burst Latency   numThreads=12   kLatencyMeasures=200M -----
##### CRTurnQueue #####
Aggregating delays for 200 million measurements...
Sorting delays...
Enqueue delay (us): 50%=1  70%=1  80%=1  90%=1  99%=2  99.9%=5  99.99%=463  99.999%=42595  max=196595
Dequeue delay (us): 50%=1  70%=1  80%=1  90%=1  99%=5  99.9%=12  99.99%=7618  99.999%=30333  max=168905
Enqueue delay (us):
50, 1
90, 1
99, 2
99.9, 5
99.99, 463
99.999, 42595
Dequeue delay (us):
50, 1
90, 1
99, 5
99.9, 12
99.99, 7618
99.999, 30333

----- Burst Latency   numThreads=16   kLatencyMeasures=200M -----
##### CRTurnQueue #####
Aggregating delays for 200 million measurements...
Sorting delays...
Enqueue delay (us): 50%=1  70%=1  80%=1  90%=1  99%=2  99.9%=5  99.99%=413  99.999%=56983  max=174279
Dequeue delay (us): 50%=1  70%=1  80%=1  90%=2  99%=5  99.9%=11  99.99%=10803  99.999%=34489  max=126404
Enqueue delay (us):
50, 1
90, 1
99, 2
99.9, 5
99.99, 413
99.999, 56983
Dequeue delay (us):
50, 1
90, 2
99, 5
99.9, 11
99.99, 10803
99.999, 34489

----- Burst Latency   numThreads=20   kLatencyMeasures=200M -----
##### CRTurnQueue #####
Aggregating delays for 200 million measurements...
Sorting delays...
Enqueue delay (us): 50%=1  70%=1  80%=1  90%=1  99%=2  99.9%=5  99.99%=590  99.999%=80746  max=290728
Dequeue delay (us): 50%=1  70%=1  80%=1  90%=2  99%=5  99.9%=26  99.99%=14994  99.999%=46232  max=273148
Enqueue delay (us):
50, 1
90, 1
99, 2
99.9, 5
99.99, 590
99.999, 80746
Dequeue delay (us):
50, 1
90, 2
99, 5
99.9, 26
99.99, 14994
99.999, 46232

----- Burst Latency   numThreads=24   kLatencyMeasures=200M -----
##### CRTurnQueue #####
Aggregating delays for 200 million measurements...
Sorting delays...
Enqueue delay (us): 50%=1  70%=1  80%=1  90%=1  99%=2  99.9%=6  99.99%=745  99.999%=99559  max=298680
Dequeue delay (us): 50%=1  70%=1  80%=1  90%=2  99%=5  99.9%=46  99.99%=19117  99.999%=52815  max=238008
Enqueue delay (us):
50, 1
90, 1
99, 2
99.9, 6
99.99, 745
99.999, 99559
Dequeue delay (us):
50, 1
90, 2
99, 5
99.9, 46
99.99, 19117
99.999, 52815

----- Burst Latency   numThreads=28   kLatencyMeasures=200M -----
##### CRTurnQueue #####
Aggregating delays for 200 million measurements...
Sorting delays...
Enqueue delay (us): 50%=1  70%=1  80%=1  90%=1  99%=2  99.9%=5  99.99%=562  99.999%=108641  max=272991
Dequeue delay (us): 50%=1  70%=1  80%=1  90%=2  99%=5  99.9%=32  99.99%=22118  99.999%=53696  max=208073
Enqueue delay (us):
50, 1
90, 1
99, 2
99.9, 5
99.99, 562
99.999, 108641
Dequeue delay (us):
50, 1
90, 2
99, 5
99.9, 32
99.99, 22118
99.999, 53696

----- Burst Latency   numThreads=30   kLatencyMeasures=200M -----
##### CRTurnQueue #####
Aggregating delays for 200 million measurements...
Sorting delays...
Enqueue delay (us): 50%=1  70%=1  80%=1  90%=1  99%=2  99.9%=5  99.99%=611  99.999%=115282  max=327969
Dequeue delay (us): 50%=1  70%=1  80%=1  90%=2  99%=5  99.9%=33  99.99%=24245  99.999%=60260  max=234501
Enqueue delay (us):
50, 1
90, 1
99, 2
99.9, 5
99.99, 611
99.999, 115282
Dequeue delay (us):
50, 1
90, 2
99, 5
99.9, 33
99.99, 24245
99.999, 60260

----- Burst Latency   numThreads=32   kLatencyMeasures=200M -----
##### CRTurnQueue #####
Aggregating delays for 200 million measurements...
Sorting delays...
Enqueue delay (us): 50%=1  70%=1  80%=1  90%=1  99%=2  99.9%=5  99.99%=511  99.999%=117147  max=300432
Dequeue delay (us): 50%=1  70%=1  80%=1  90%=2  99%=5  99.9%=27  99.99%=24961  99.999%=58434  max=236729
Enqueue delay (us):
50, 1
90, 1
99, 2
99.9, 5
99.99, 511
99.999, 117147
Dequeue delay (us):
50, 1
90, 2
99, 5
99.9, 27
99.99, 24961
99.999, 58434
```

**Comment by KyleMayes** [comment]
The latencies in the graphs from the paper don't start diverging significantly until you pass 4 threads accessing the queue simultaneously.

Their benchmarks were on a 32-core server, and my machine has two cores and it seems yours does too (though being an i7 it does have HyperThreading, so 4 virtual cores). I wouldn't expect either of us to be able to replicate their benchmarks with these machines.

I'll go ahead and publish this crate, I'm pretty confident that it works, I was just wary of publishing it initially when I was unable to confirm for sure that it accomplishes what it is supposed to.

**Comment by KyleMayes** [comment]
Published as `v0.1.0`.
